api/query/atoms.go

```diff
@@ -15,61 +15,118 @@ import (
 
 var browsers = shared.GetDefaultBrowserNames()
 
-type query interface {
-	toPlan(runIDs []int64) plan
+// AbstractQuery is an intermetidate representation of a test results query that
+//  has not been bound to specific shared.TestRun specs for processing.
+type AbstractQuery interface {
+	BindToRuns(runs []shared.TestRun) ConcreteQuery
 }
 
-type runQuery struct {
+// RunQuery is the internal representation of a query recieved from an HTTP
+// client, including the IDs of the test runs to query, and the structured query
+// to run.
+type RunQuery struct {
 	runIDs []int64
-	query
+	AbstractQuery
 }
 
-func (rq runQuery) toPlan() plan {
-	return rq.query.toPlan(rq.runIDs)
+// TestNamePattern is a query atom that matches test names to a pattern string.
+type TestNamePattern struct {
+	Pattern string
 }
 
-type testNamePattern struct {
-	pattern string
-}
-
-func (tnp testNamePattern) toPlan(runIDs []int64) plan {
-	return nil
+// BindToRuns for TestNamePattern is a no-op: TestNamePattern implements both
+// AbstractQuery and ConcreteQuery because it is independent of test runs.
+func (tnp TestNamePattern) BindToRuns(runs []shared.TestRun) ConcreteQuery {
+	return tnp
 }
 
 type testStatusConstraint struct {
 	browserName string
 	status      int64
 }
 
-func (tsc testStatusConstraint) toPlan(runIDs []int64) plan {
-	return nil
+func (tsc testStatusConstraint) BindToRuns(runs []shared.TestRun) ConcreteQuery {
+	ids := make([]int64, 0, len(runs))
+	for _, run := range runs {
+		if run.BrowserName == tsc.browserName {
+			ids = append(ids, run.ID)
+		}
+	}
+	if len(ids) == 0 {
+		return True{}
+	}
+	if len(ids) == 1 {
+		return RunTestStatusConstraint{ids[0], tsc.status}
+	}
+
+	q := Or{make([]ConcreteQuery, len(ids))}
+	for i := range ids {
+		q.Args[i] = RunTestStatusConstraint{ids[i], tsc.status}
+	}
+	return q
 }
 
 type not struct {
-	not query
+	not AbstractQuery
 }
 
-func (n not) toPlan(runIDs []int64) plan {
-	return nil
+func (n not) BindToRuns(runs []shared.TestRun) ConcreteQuery {
+	return Not{n.not.BindToRuns(runs)}
 }
 
 type or struct {
-	or []query
+	or []AbstractQuery
 }
 
-func (o or) toPlan(runIDs []int64) plan {
-	return nil
+func (o or) BindToRuns(runs []shared.TestRun) ConcreteQuery {
+	q := Or{make([]ConcreteQuery, 0, len(o.or))}
+	for i := range o.or {
+		sub := o.or[i].BindToRuns(runs)
+		if _, ok := sub.(True); ok {
+			return True{}
+		}
+		if _, ok := sub.(False); ok {
+			continue
+		}
+		q.Args = append(q.Args, sub)
+	}
+	if len(q.Args) == 0 {
+		return True{}
+	}
+	if len(q.Args) == 1 {
+		return q.Args[0]
+	}
+	return q
 }
 
 type and struct {
-	and []query
+	and []AbstractQuery
 }
 
-func (a and) toPlan(runIDs []int64) plan {
-	return nil
+func (a and) BindToRuns(runs []shared.TestRun) ConcreteQuery {
+	q := And{make([]ConcreteQuery, 0, len(a.and))}
```

Comment:
I feel this is a bit hard to read; can you name the field in the `And` struct that you're make-ing?

Reply:
Sure. We may not see it (see `len()` guards at end of method), so I'll just make the slice here and apply it (with field name) in the case that we return an `And`.